src/Perlang.Tests.Integration/Typing/TypingTests.cs

```diff
@@ -124,6 +124,53 @@ public void var_declaration_with_initializer_correctly_infers_type_from_assignme
             Assert.Equal("System.Int32", result);
         }
 
+        [Fact]
+        public void var_declaration_with_initializer_detects_inference_attempt_from_nil()
+        {
+            // The following program should fail, since there is no way the compiler can infer any type information from
+            // a 'nil' initializer.
+            string source = @"
+                var s = nil;
+
+                print s;
+            ";
+
+            var result = EvalWithValidationErrorCatch(source);
+            var exception = result.Errors.FirstOrDefault();
+
+            Assert.Single(result.Errors);
+            Assert.Matches("Cannot assign nil to an implicitly typed local variable", exception.Message);
+        }
```

Comment:
This added test validates the new check in `TypesResolvedValidator`, added in this PR.